vac/raw/decentralized-messaging-ethereum.md

```diff
@@ -0,0 +1,737 @@
+
+
+### Post-compromise security updates and group member removal
+
+The functions `update` and `remove` share similarities with `create`: 
+they both call the function `generate-seed` to encrypt a new seed secret for each group member. 
+The distinction lies in the determination of the group members using `member-view`. 
+In the case of `remove`, the user being removed is excluded from the recipients of the seed secret. 
+Additionally, the control message they construct is designated with type *update* or *remove* respectively.
+
+Likewise, `process-update` and `process-remove` are akin to `process-create`. 
+The function `process-update` skips the update of gamma.history, 
+whereas `process-remove` includes a removal operation in the history.
+
+#### update
+```
+control = ("update", ++gamma.mySeq, empty_string)
+recipients = member-view(gamma, gamma.myId) - {gamma.myId}
+(gamma, dmsgs) = generate-seed(gamma, recipients)
+(gamma, _, _, I , _) = process-update(gamma, gamma.myId, gamma.mySeq, empty_string, empty_string)
+return (gamma, control, dmsgs, I)
+```
+
+#### remove
+```
+control = ("remove", ++gamma.mySeq, empty)
+recipients = member-view(gamma, gamma.myId) - {ID, gamma.myId}
+(gamma, dmsgs) = generate-seed(gamma, recipients)
+(gamma, _, _, I , _) = process-update(gamma, gamma.myId, gamma.mySeq, ID, empty_string)
+return (gamma, control, dmsgs, I)
+```
+
+#### process-update
+`return process-seed(gamma, sender, seq, dmsg)`
+
+#### process-remove
+```
+op = ("remove", sender, seq, removed)
+gamma.history = gamma.history + {op}
+return process-seed(gamma, sender, seq, dmsg)
+```
+
+### Group member addition
+
+#### add
+When adding a new group member, 
+an existing member initiates the process by invoking the `add` function and 
+providing the ID of the user to be added.
+This function prepares a control message marked as *add* for broadcast to the group. 
+Simultaneously, it creates a welcome message intended for the new member as a direct message. 
+This *welcome* message includes the current state of the sender's KDF ratchet, 
+encrypted using `2SM`, along with the history of group membership operations conducted so far.
+
+```
+control = ("add", ++gamma.mySeq, ID)
+(gamma, c) = encrypt-to(gamma, ID, gamma.ratchet[gamma.myId])
+op = ("add", gamma.myId, gamma.mySeq, ID)
+welcome = (gamma.history + {op}, c)
+(gamma, _, _, $I$, _) = process-add(gamma, gamma.myId, gamma.mySeq, ID, empty_string)
+return (gamma, control, {(ID, welcome)}, $I$)
+```
+
+#### process-add
+This function is invoked by both the sender and 
+each recipient of an *add* message, which includes the new group member. 
+If the local user is the newly added member, 
+the function proceeds to call `process-welcome` and then exits. 
+Otherwise, it extends `gamma.history` with the `add` operation.
+
+Line 5 determines whether the local user was already a group member at the time the "add" message was sent; 
```

Comment:
Here and below in this section is a line by line description and perhaps then we should add them to the pseudo code?